Source/DafnyStandardLibraries/src/dafnystdlibs/Wrappers.md

```diff
@@ -0,0 +1,191 @@
+
+## The `Wrappers` module
+
+The Wrappers module holds some
+simple datatypes to support common patterns, such as optional values or the result of operations that can fail.
+These are additionally useful with Dafny's abrupt-termination-on-failure `:-` operator.
+
+Any user datatype can serve this purpose, as long as it has an `IsFailure?` predicate 
+(in which case it is known as a [_failure-compatible_](https://dafny.org/latest/DafnyRef/DafnyRef#sec-failure-compatible-types) type --- an FC-type). 
+
+In this library module Dafny defines three such types:
+- `Option<R>` - which is either `Some` with a value of type `T` or a `None` with no information
+- `Outcome<E>` - which is either `Pass` with no informatino or `Fail` with an error value of type `E` (often a `string`) 
+- `Result<R,E>` - which is either `Success` with a value of type `R` or `Failure` with error value of type `E`
+
+These are common programming idioms. The main complication comes when they are mixed in the same program.
+
+### Option
+
+Consider this routine that looks for a value in a sequence, beginning at position `k`, returning its index:
+<!-- %check-verify %save tmp-find.dfy -->
+```dafny
+import opened DafnyStdLibs.Wrappers
+function Find<T(==)>(s: seq<T>, k: int, value: T): (r: Option<int>)
+  requires 0 <= k <= |s|
+  ensures  r.Some? ==> 0 <= r.Extract() < |s| && s[r.Extract()] == value
+  ensures  r.None? ==> forall i | k <= i < |s| :: s[i] != value
+  decreases |s| - k
+{
+  if k >= |s| then None else if s[k] == value then Some(k) else Find(s, k+1, value)
+}
+```
+
+It could be used in a method like this
+
+<!-- %check-resolve %use tmp-find.dfy -->
+```dafny
+method m(s: seq<int>) returns (r: Option<int>) {
+  var value: int;
+  // do various calculations
+  var index: int :- Find(s, 0, value);
+  // more calculations
+  return Some(value);
+}
+```
+
+You could just capture the result of `Find` using `:=` in a `Option<int>` variable and inspect it. But if the `None` condition is 
+generally a rare error, it is easy to forget to always check that each operation was successful. Instead, the `:-` changes the 
```

Comment:
I think I prefer this text not to have line breaks, so the viewer can take care of the appropriate breaking.

Reply:
My general preference is https://sembr.org/ personally. I didn't want to edit this too heavily so we can see a clean diff against the original, though: https://github.com/dafny-lang/libraries/blob/master/src/dafny/Wrappers.md

Reply:
I don't get it. Why wouldn't you want regular text to look the same when writing as when reading? If we're soft-wrapping it with no explicit line breaks when reading, why not also write it like that?